docs/Environment.md
# Environment Variable

| Env | |
| --- | -- |
| `CONSOLE_MINIO_SERVER` | "http://localhost:9000" |
| `CONSOLE_MINIO_REGION` | "us-east-1" |
| `CONSOLE_HOSTNAME` | "" |
| `CONSOLE_PORT` | 9090 |
| `CONSOLE_TLS_PORT` | 9443 |
| `CONSOLE_SUBPATH` | i.e. /console |
| `CONSOLE_DEBUG_LOGLEVEL` | 0 - 6 |
| `CONSOLE_SHARE_MINIO_URL` | "off"
| `CONSOLE_SECURE_ALLOWED_HOSTS` | "" |
| `CONSOLE_SECURE_ALLOWED_HOSTS_ARE_REGEX` | "off" |
| `CONSOLE_SECURE_FRAME_DENY` | "on" |
| `CONSOLE_SECURE_CONTENT_TYPE_NO_SNIFF` | "on" |
| `CONSOLE_SECURE_BROWSER_XSS_FILTER` | "on" |
| `CONSOLE_SECURE_CONTENT_SECURITY_POLICY` | "" |
| `CONSOLE_SECURE_CONTENT_SECURITY_POLICY_REPORT_ONLY` | "" |
| `CONSOLE_SECURE_HOSTS_PROXY_HEADERS` | "" |
| `CONSOLE_SECURE_STS_SECONDS` | 0 |
| `CONSOLE_SECURE_STS_INCLUDE_SUB_DOMAINS` | "off" |
| `CONSOLE_SECURE_STS_PRELOAD` | "off" |
| `CONSOLE_SECURE_TLS_REDIRECT` | "off" |
| `CONSOLE_SECURE_TLS_HOST` | "" |
| `CONSOLE_SECURE_TLS_TEMPORARY_REDIRECT` | "off" |
| `CONSOLE_SECURE_FORCE_STS_HEADER` | "off" |
| `CONSOLE_SECURE_PUBLIC_KEY` | |
| `CONSOLE_SECURE_REFERRER_POLICY` | "" |
| `CONSOLE_SECURE_FEATURE_POLICY` | "" |
| `CONSOLE_SECURE_EXPECT_CT_HEADER` | |
| `CONSOLE_PROMETHEUS_URL` | |
| `CONSOLE_PROMETHEUS_AUTH_TOKEN` | |
| `CONSOLE_PROMETHEUS_AUTH_USERNAME` | |
| `CONSOLE_PROMETHEUS_AUTH_PASSWORD` | |
| `CONSOLE_PROMETHEUS_JOB_ID` | "minio-job" |
| `CONSOLE_PROMETHEUS_EXTRA_LABELS` | |
| `CONSOLE_LOG_QUERY_URL` | |
| `CONSOLE_LOG_QUERY_AUTH_TOKEN` | "" |
| `CONSOLE_MAX_CONCURRENT_UPLOADS` | "10" |
| `CONSOLE_MAX_CONCURRENT_DOWNLOADS` | "20" |
| `CONSOLE_DEV_MODE` | "off" |
| `CONSOLE_BROWSER_REDIRECT_URL` | |
| `LOGSEARCH_QUERY_AUTH_TOKEN` | |
| `CONSOLE_IDP_DISPLAY_NAME` | `MINIO_IDENTITY_OPENID_DISPLAY_NAME` |
| `CONSOLE_IDP_URL` | `MINIO_IDENTITY_OPENID_CONFIG_URL` |
| `CONSOLE_IDP_CLIENT_ID` | `MINIO_IDENTITY_OPENID_CLIENT_ID` | 
| `CONSOLE_IDP_SECRET` | `MINIO_IDENTITY_OPENID_CLIENT_SECRET` | 
| `CONSOLE_IDP_CALLBACK` | `MINIO_BROWSER_REDIRECT_URL` | 
| `CONSOLE_IDP_CALLBACK_DYNAMIC` | `MINIO_IDENTITY_OPENID_REDIRECT_URI_DYNAMIC` |
| `CONSOLE_IDP_SCOPES` | `MINIO_IDENTITY_OPENID_SCOPES` | 
| `CONSOLE_IDP_USERINFO` | `MINIO_IDENTITY_OPENID_CLAIM_USERINFO` | 
| `CONSOLE_IDP_ROLE_ARN` | | 
| `CONSOLE_IDP_END_SESSION_ENDPOINT` | |
| `CONSOLE_LDAP_ENABLED` | |
| `CONSOLE_STS_DURATION` | time.Duration format, ie: 3600s, 2h45m, 1h, etc
| `CONSOLE_PBKDF_PASSPHRASE` | |
| `CONSOLE_PBKDF_SALT` | |
| `CONSOLE_LOGGER_JSON_ENABLE` | 
| `CONSOLE_LOGGER_ANONYMOUS_ENABLE` |
| `CONSOLE_LOGGER_QUIET_ENABLE` |
| `CONSOLE_GLOBAL_DEPLOYMENT_ID` |
| `CONSOLE_LOGGER_WEBHOOK_ENABLE` |
| `CONSOLE_LOGGER_WEBHOOK_ENDPOINT` |
| `CONSOLE_LOGGER_WEBHOOK_AUTH_TOKEN` |
| `CONSOLE_LOGGER_WEBHOOK_CLIENT_CERT` |
| `CONSOLE_LOGGER_WEBHOOK_CLIENT_KEY` |
| `CONSOLE_LOGGER_WEBHOOK_QUEUE_SIZE` |
| `CONSOLE_AUDIT_WEBHOOK_ENABLE` |
| `CONSOLE_AUDIT_WEBHOOK_ENDPOINT` |
| `CONSOLE_AUDIT_WEBHOOK_AUTH_TOKEN` |
| `CONSOLE_AUDIT_WEBHOOK_CLIENT_CERT` |
| `CONSOLE_AUDIT_WEBHOOK_CLIENT_KEY` |
| `CONSOLE_AUDIT_WEBHOOK_QUEUE_SIZE` |


CHANGELOG.md
# Changelog

## Release v1.9.1

Bug Fix:

- Updated project dependencies
- Updated go version from 1.24.10 to 1.24.11 to fix Security vulnerability

## Release v1.9.0

Breaking Change:

- ODIC: `CONSOLE_IDP_CALLBACK` and `MINIO_BROWSER_REDIRECT_URL` now expect the Console URL without `/oauth_callback` at the end

Features:

- Supports Prometheus basic auth
- ReadOnly and disabled feature for CodeEditor, SpeedtestResult Json
- Adds View to see Health Info Report Results as JSON Preview
- New SSO URL `/sso` for auto redirect to OIDC Provider
- Shows and option to load more than 20 versions
- Login page shows an indicator that LDAP is enabled
- Use Quota Size field instead of the deprecated Quota field
- Console container now runs rootless as user 1000:1000
- Show console package version on license page

Bug Fix:

- Some OIDC confussion around ROLE_POLICY vs. ROLE_ARN
- Fix download option in file preview
- Set goreleaser bindir for linux packages to /usr/local/bin
- Fix tag retrieval in ObjectDetailPanel component
- Fix metrics display for objects sizes between 1024B and 1MB

Additional Changes:

- Alot of dependencies updates

## Release v1.8.1

Release focuses on debranding by dropping **MinIO®** from names and logos

Features:
- OIDC SSO Login support see [docs](./docs/OIDC.md)
- Self-Update of binarys over Github Releases with `./console update`

Deprecations:
- Deprecates CONSOLE_ANIMATED_LOGIN animated Login video background 
- Deprecates Inclusion of sourcemaps in Prod Releases of Web-App

Build:
- web-app frontend build size 28 MB down to 9 MB
- reduced binary size ~60 MB to ~45 MB

Pictures see releases 

## Release v1.8.0

This release is bringing back long-deprecated features:

- Undeprecated Lifecyle and Tierung UI (minio#3470)
- Undeprecated Site Replication in UI (minio#3469)
- Unremoved Tools support (minio#3467)
    - Health Info
    - Speedtest
    - Profiling
    - Inspect
    - Trace
    - Watch
- Removed Subnet, Registration, Call Home Support again after Revert
- Small License and Login Page updated

## Release v1.7.8

Bug Fix:

- Fixed Dependencies vulnerabilities + updated Dependencies
- Allow console to recognize s3.Delete*
- Fix regex pattern in webhook management
- Fix golangci-lint issues 
- Decreased Browser direct download threshold to 5GiB

## Release v1.7.7

There are actually no changes compared to v1.7.6; I'm just getting the release and builds ready.
- Binary Releases
- Packages
- Container Builds

See Releases

## Release v1.7.6

Bug Fix:

- Fix null pointer exception in Admin Info
- Ignore leading or trailing spaces in login request
- Fix file path on drag and drop
- Fix typo in User DN Search Filter example

## Release v1.7.5

Bug Fix:

- Fixed leaks during ZIP multiobject downloads
- Allow spaces in Policy names

## Release v1.7.4

Deprecations:

- Deprecated support tools User Interface in favor of mc admin commands. Please refer to the [MinIO Client documentation page](https://docs.min.io/community/minio-object-store/reference/minio-mc.html) for more information.
- Deprecated Site replication User Interface in favor of mc admin commands. Please refer to the [MinIO Site Replication page](https://docs.min.io/community/minio-object-store/reference/minio-mc-admin/mc-admin-replicate.html) for more information.
- Deprecated Lifecycle & Tiers User Interface in favor of mc admin commands. Please refer to the [MinIO Tiers page](https://docs.min.io/community/minio-object-store/reference/minio-mc/mc-ilm-tier.html) for more information.

Bug Fix:

- Avoid loading unpkg.com call when login animation is off

## Release v1.7.3

Bug Fix:

- Use a fixed public license verification key
- Show non-expiring access keys as `no-expiry` instead of Jan 1, 1970
- Use "join Slack" button for non-commercial edition instead of "Signup"
- Fix setting policies on groups that have spaces

## Release v1.7.2

Bug Fix:

- Fixed issue in Server Health Info
- Fixed Security vulnerability in dependencies
- Fixed client string in trace message

Additional Changes:

- Remove live logs in Call Home Page
- Update License page

## Release v1.7.1

Bug Fix:

- Fixed issue that could cause a failure when attempting to view deleted files in the object browser
- Return network error when logging in and the network connection fails

Additional Changes:

- Added debug logging for console HTTP request (see [PR #3440](https://github.com/minio/console/pull/3440) for more detailed information)

## Release v1.7.0

Bug Fix:

- Fixed directory listing
- Fix MinIO videos link

Additional Changes:

- Removed deprecated KES functionality

## Release v1.6.3

Additional Changes:

- Updated go.mod version

## Release v1.6.2

Bug Fix:

- Fixed minor user session issues
- Updated project dependencies

Additional Changes:

- Improved Drives List visualization
- Improved WS request logic
- Updated License page with current MinIO plans.

## Release v1.6.1

Bug Fix:

- Fixed objectManager issues under certain conditions
- Fixed Security vulnerability in dependencies

Additional Changes:

- Improved Share Link behavior

## Release v1.6.0

Bug Fix:

- Fixed share link encoding
- Fixed Edit Lifecycle Storage Class
- Added Tiers Improvements for Bucket Lifecycle management

Additional Changes:

- Vulnerability updates
- Update Logo logic

## Release v1.5.0

Features:

- Added remove Tier functionality

Bug Fix:

- Fixed ILM rule tags not being shown
- Fixed race condition Object Browser websocket
- Fixed Encryption page crashing on empty response
- Fixed Replication Delete Marker comparisons

Additional Changes:

- Use automatic URI encoding for APIs
- Vulnerability updates

## Release v1.4.0

Features:

- Added VersionID support to metadata details
- Improved Websockets handlers

Bug Fix:

- Fixed vulnerabilities and updated dependencies
- Fixed an issue with Download URL decoding
- Fixed leak in Object Browser Websocket
- Minor UX fixes

## Release v1.3.0

Features:


api/config.go
// This file is part of MinIO Console Server
// Copyright (c) 2021 MinIO, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package api

import (
	"crypto/tls"
	"crypto/x509"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/minio/console/pkg/auth/idp/oauth2"
	xcerts "github.com/minio/pkg/v3/certs"
	"github.com/minio/pkg/v3/env"
	xnet "github.com/minio/pkg/v3/net"
)

var (
	// Port console default port
	Port = "9090"

	// Hostname console hostname
	// avoid listening on 0.0.0.0 by default
	// instead listen on all IPv4 and IPv6
	// - Hostname should be empty.
	Hostname = ""

	// TLSPort console tls port
	TLSPort = "9443"

	// TLSRedirect console tls redirect rule
	TLSRedirect = "on"

	ConsoleResourceName = "console-ui"
)

var (
	// GlobalRootCAs is CA root certificates, a nil value means system certs pool will be used
	GlobalRootCAs *x509.CertPool
	// GlobalPublicCerts has certificates Console will use to serve clients
	GlobalPublicCerts []*x509.Certificate
	// GlobalTLSCertsManager custom TLS Manager for SNI support
	GlobalTLSCertsManager *xcerts.Manager
	// GlobalTransport is common transport used for all HTTP calls, this is set via
	// MinIO server to be the correct transport, however we still define some defaults
	// here just in case.
	GlobalTransport = &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		DialContext: (&net.Dialer{
			Timeout:   10 * time.Second,
			KeepAlive: 15 * time.Second,
		}).DialContext,
		MaxIdleConns:          1024,
		MaxIdleConnsPerHost:   1024,
		IdleConnTimeout:       90 * time.Second,
		TLSHandshakeTimeout:   10 * time.Second,
		ExpectContinueTimeout: 10 * time.Second,
		DisableCompression:    true, // Set to avoid auto-decompression
		TLSClientConfig: &tls.Config{
			// Can't use SSLv3 because of POODLE and BEAST
			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
			// Can't use TLSv1.1 because of RC4 cipher usage
			MinVersion: tls.VersionTLS12,
			// Console runs in the same pod/node as MinIO this is acceptable.
			InsecureSkipVerify: true,
			RootCAs:            GlobalRootCAs,
		},
	}
)

// MinIOConfig represents application configuration passed in from the MinIO
// server to the console.
type MinIOConfig struct {
	OpenIDProviders oauth2.OpenIDPCfg
}

// GlobalMinIOConfig is the global application configuration passed in from the
// MinIO server.
var GlobalMinIOConfig MinIOConfig

func getMinIOServer() string {
	return strings.TrimSpace(env.Get(ConsoleMinIOServer, "http://localhost:9000"))
}

func GetMinIORegion() string {
	return strings.TrimSpace(env.Get(ConsoleMinIORegion, ""))
}

func getMinIOEndpoint() string {
	u, err := xnet.ParseHTTPURL(getMinIOServer())
	if err != nil {
		panic(err)
	}
	return u.Host
}

func getMinIOEndpointIsSecure() bool {
	u, err := xnet.ParseHTTPURL(getMinIOServer())
	if err != nil {
		panic(err)
	}
	return u.Scheme == "https"
}

// GetHostname gets console hostname set on env variable,
// default one or defined on run command
func GetHostname() string {
	return strings.ToLower(env.Get(ConsoleHostname, Hostname))
}

// GetPort gets console por set on env variable
// or default one
func GetPort() int {
	port, err := strconv.Atoi(env.Get(ConsolePort, Port))
	if err != nil {
		port = 9090
	}
	return port
}

// GetTLSPort gets console tls port set on env variable
// or default one
func GetTLSPort() int {
	port, err := strconv.Atoi(env.Get(ConsoleTLSPort, TLSPort))
	if err != nil {
		port = 9443
	}
	return port
}

// If GetTLSRedirect is set to true, then only allow HTTPS requests. Default is true.
func GetTLSRedirect() string {
	return strings.ToLower(env.Get(ConsoleSecureTLSRedirect, TLSRedirect))
}

// Get secure middleware env variable configurations
func GetSecureAllowedHosts() []string {
	allowedHosts := env.Get(ConsoleSecureAllowedHosts, "")
	if allowedHosts != "" {
		return strings.Split(allowedHosts, ",")
	}
	return []string{}
}

// AllowedHostsAreRegex determines, if the provided AllowedHosts slice contains valid regular expressions. Default is false.
func GetSecureAllowedHostsAreRegex() bool {
	return strings.ToLower(env.Get(ConsoleSecureAllowedHostsAreRegex, "off")) == "on"
}

// If FrameDeny is set to true, adds the X-Frame-Options header with the value of `DENY`. Default is true.
func GetSecureFrameDeny() bool {
	return strings.ToLower(env.Get(ConsoleSecureFrameDeny, "on")) == "on"
}

// If ContentTypeNosniff is true, adds the X-Content-Type-Options header with the value `nosniff`. Default is true.
func GetSecureContentTypeNonSniff() bool {
	return strings.ToLower(env.Get(ConsoleSecureContentTypeNoSniff, "on")) == "on"
}

// If BrowserXssFilter is true, adds the X-XSS-Protection header with the value `1; mode=block`. Default is true.
func GetSecureBrowserXSSFilter() bool {
	return strings.ToLower(env.Get(ConsoleSecureBrowserXSSFilter, "on")) == "on"
}

// ContentSecurityPolicy allows the Content-Security-Policy header value to be set with a custom value. Default is "".
// Passing a template string will replace `$NONCE` with a dynamic nonce value of 16 bytes for each request which can be
// later retrieved using the Nonce function.
func GetSecureContentSecurityPolicy() string {
	return env.Get(ConsoleSecureContentSecurityPolicy, "")
}

// ContentSecurityPolicyReportOnly allows the Content-Security-Policy-Report-Only header value to be set with a custom value. Default is "".
func GetSecureContentSecurityPolicyReportOnly() string {
	return env.Get(ConsoleSecureContentSecurityPolicyReportOnly, "")
}

// HostsProxyHeaders is a set of header keys that may hold a proxied hostname value for the request.
func GetSecureHostsProxyHeaders() []string {
	allowedHosts := env.Get(ConsoleSecureHostsProxyHeaders, "")
	if allowedHosts != "" {
		return strings.Split(allowedHosts, ",")
	}
	return []string{}
}

// TLSHost is the host name that is used to redirect HTTP requests to HTTPS. Default is "", which indicates to use the same host.
func GetSecureTLSHost() string {
	tlsHost := env.Get(ConsoleSecureTLSHost, "")
	if tlsHost == "" && Hostname != "" {
		return net.JoinHostPort(Hostname, TLSPort)
	}
	return ""
}

// STSSeconds is the max-age of the Strict-Transport-Security header. Default is 0, which would NOT include the header.
func GetSecureSTSSeconds() int64 {
	seconds, err := strconv.Atoi(env.Get(ConsoleSecureSTSSeconds, "0"))
	if err != nil {
		seconds = 0
	}
	return int64(seconds)
}

// If STSIncludeSubdomains is set to true, the `includeSubdomains` will be appended to the Strict-Transport-Security header. Default is false.
func GetSecureSTSIncludeSubdomains() bool {
	return strings.ToLower(env.Get(ConsoleSecureSTSIncludeSubdomains, "off")) == "on"
}

// If STSPreload is set to true, the `preload` flag will be appended to the Strict-Transport-Security header. Default is false.
func GetSecureSTSPreload() bool {
	return strings.ToLower(env.Get(ConsoleSecureSTSPreload, "off")) == "on"
}

// If TLSTemporaryRedirect is true, the a 302 will be used while redirecting. Default is false (301).
func GetSecureTLSTemporaryRedirect() bool {
	return strings.ToLower(env.Get(ConsoleSecureTLSTemporaryRedirect, "off")) == "on"
}

// STS header is only included when the connection is HTTPS.
func GetSecureForceSTSHeader() bool {
	return strings.ToLower(env.Get(ConsoleSecureForceSTSHeader, "off")) == "on"
}

// ReferrerPolicy allows the Referrer-Policy header with the value to be set with a custom value. Default is "".
func GetSecureReferrerPolicy() string {
	return env.Get(ConsoleSecureReferrerPolicy, "")
}

// FeaturePolicy allows the Feature-Policy header with the value to be set with a custom value. Default is "".
func GetSecureFeaturePolicy() string {
	return env.Get(ConsoleSecureFeaturePolicy, "")
}

func getLogSearchAPIToken() string {
	if v := env.Get(ConsoleLogQueryAuthToken, ""); v != "" {
		return v
	}
	return env.Get(LogSearchQueryAuthToken, "")
}

func getLogSearchURL() string {
	return env.Get(ConsoleLogQueryURL, "")
}

func getPrometheusURL() string {
	return env.Get(PrometheusURL, "")
}

func getPrometheusAuthToken() string {
	return env.Get(PrometheusAuthToken, "")
}

func getPrometheusAuthUserPass() (string, string, bool) {
	username := env.Get(PrometheusAuthUsername, "")
	password := env.Get(PrometheusAuthPassword, "")
	if username == "" && password == "" {
		return "", "", false
	}

	return username, password, true
}

func getPrometheusJobID() string {
	return env.Get(PrometheusJobID, "minio-job")
}

func getPrometheusExtraLabels() string {
	return env.Get(PrometheusExtraLabels, "")
}

func getMaxConcurrentUploadsLimit() int64 {
	cu, err := strconv.ParseInt(env.Get(ConsoleMaxConcurrentUploads, "10"), 10, 64)
	if err != nil {
		return 10
	}

	return cu
}

func getMaxConcurrentDownloadsLimit() int64 {
	cu, err := strconv.ParseInt(env.Get(ConsoleMaxConcurrentDownloads, "20"), 10, 64)
	if err != nil {
		return 20
	}

	return cu
}

func getConsoleDevMode() bool {
	return strings.ToLower(env.Get(ConsoleDevMode, "off")) == "on"
}

func getConsoleBrowserRedirectURL() string {
	return env.Get(ConsoleBrowserRedirectURL, "")
}

func getConsoleShareMinIOURL() bool {
	return strings.ToLower(env.Get(ConsoleShareMinIOURL, "off")) == "on"
}

func BuildOpenIDConsoleConfig() oauth2.OpenIDPCfg {
	pcfg := map[string]oauth2.ProviderConfig{}

	url := env.Get(ConsoleIDPURL, env.Get(MinioIdentifyOpenIDConfigURL, ""))
	clientID := env.Get(ConsoleIDPClientID, env.Get(MinioIdentifyOpenIDClientID, ""))
	clientSecret := env.Get(ConsoleIDPSecret, env.Get(MinioIdentifyOpenIDClientSecret, ""))
	redirectCallback := env.Get(ConsoleIDPCallbackURL, env.Get(MinioBrowserRedirectURL, ""))

	// Only set config if url, clientID, clientSecret and redirectCallback are provided
	if url != "" && clientID != "" && clientSecret != "" && redirectCallback != "" {
		pcfg = map[string]oauth2.ProviderConfig{
			"OIDC": {
				URL:                     url,
				ClientID:                clientID,
				ClientSecret:            clientSecret,
				RedirectCallback:        redirectCallback + "/oauth_callback",
				DisplayName:             env.Get(ConsoleIDPDisplayName, env.Get(MinioIdentifyOpenIDDisplayName, "")),
				Scopes:                  env.Get(ConsoleIDPScopes, env.Get(MinioIdentifyOpenIDScopes, "openid,profile,email")),
				Userinfo:                env.Get(ConsoleIDPUserInfo, env.Get(MinioIdentifyOpenIDClaimUserinfo, "")) == "on",
				RedirectCallbackDynamic: env.Get(ConsoleIDPCallbackURLDynamic, env.Get(MinioIdentifyOpenIDRedirectURIDynamic, "")) == "on",
				RoleArn:                 env.Get(ConsoleIDPRoleArn, ""),
				EndSessionEndpoint:      env.Get(ConsoleIDPEndSessionEndpoint, ""),
			},
		}
	}

	return pcfg
}
